llrt_core/src/custom_resolver.rs

```diff
@@ -0,0 +1,570 @@
+    if Path::new(&file).is_file() {
+        trace!("|  load_index(2): {}", file);
+        return Ok(Some(file));
+    }
+
+    // 3. If X/index.lrt is a file, load X/index.lrt as JavaScript bytecode. STOP
+    let file = [x, "/index.lrt"].concat();
+    if Path::new(&file).is_file() {
+        trace!("|  load_index(3): {}", file);
+        return Ok(Some(file));
+    }
+
+    Ok(None)
+}
+
+// LOAD_AS_DIRECTORY(X)
+fn load_as_directory(ctx: &Ctx<'_>, x: &str) -> Result<Option<String>> {
+    trace!("|  load_as_directory(x): {}", x);
+
+    // 1. If X/package.json is a file,
+    let file = [x, "/package.json"].concat();
+    if Path::new(&file).is_file() {
+        // a. Parse X/package.json, and look for "main" field.
+        // b. If "main" is a falsy value, GOTO 2.
+        if let Some(main) = get_main_field(ctx, &file)? {
+            // c. let M = X + (json main field)
+            let m = [x, "/", &main].concat();
+            // d. LOAD_AS_FILE(M)
+            if let Ok(Some(path)) = load_as_file(ctx, &m) {
+                trace!("|  load_as_directory(1.d): {}", path);
+                return Ok(Some(path));
+            }
+            // e. LOAD_INDEX(M)
+            if let Ok(Some(path)) = load_index(ctx, &m) {
+                trace!("|  load_as_directory(1.e): {}", path);
+                return Ok(Some(path));
+            }
+            // f. LOAD_INDEX(X) DEPRECATED
+
+            // g. THROW "not found"
+            return Err(Error::new_resolving("", x.to_string()));
+        }
+    }
+
+    // 2. LOAD_INDEX(X)
+    if let Ok(Some(path)) = load_index(ctx, x) {
+        trace!("|  load_as_directory(2): {}", path);
+        return Ok(Some(path));
+    }
+
+    Ok(None)
+}
+
+// LOAD_NODE_MODULES(X, START)
+fn load_node_modules(ctx: &Ctx<'_>, x: &str, start: &str, is_esm: bool) -> Option<String> {
+    trace!("|  load_node_modules(x, start): ({}, {})", x, start);
+
+    // 1. let DIRS = NODE_MODULES_PATHS(START)
+    let mut cache = NODE_MODULES_PATHS_CACHE.lock().unwrap();
+    let dirs = cache
+        .entry(start.to_string())
+        .or_insert_with(|| node_modules_paths(start));
+
+    // 2. for each DIR in DIRS:
+    for dir in dirs {
+        // a. LOAD_PACKAGE_EXPORTS(X, DIR)
+        if let Ok(path) = load_package_exports(ctx, x, dir, is_esm) {
+            trace!("|  load_node_modules(2.a): {}", path);
+            return Some(path);
+        }
+        let dir_slash_x = [dir, "/", x].concat();
+        // b. LOAD_AS_FILE(DIR/X)
+        if let Ok(Some(path)) = load_as_file(ctx, &dir_slash_x) {
+            trace!("|  load_node_modules(2.b): {}", path);
+            return Some(path);
+        }
+        // c. LOAD_AS_DIRECTORY(DIR/X)
+        if let Ok(Some(path)) = load_as_directory(ctx, &dir_slash_x) {
+            trace!("|  load_node_modules(2.c): {}", path);
+            return Some(path);
+        }
+    }
+
+    None
+}
+
+// NODE_MODULES_PATHS(START)
+fn node_modules_paths(start: &str) -> Vec<String> {
```

Comment:
Can we do something like this? (untested)
```rust
fn node_modules_paths(start: &str) -> Vec<PathBuf> {
    let path = Path::new(start);
    let mut dirs = Vec::new();
    let mut current = Some(path);

    // Iterate through parent directories
    while let Some(dir) = current {
        if dir.file_name().map_or(false, |name| name != "node_modules") {
            let mut node_modules = dir.to_path_buf();
            node_modules.push("node_modules");
            dirs.push(node_modules);
        }
        current = dir.parent();
    }

    // Add global folders
    if let Some(home) = home::home_dir() {
        dirs.push(home.join(".node_modules"));
        dirs.push(home.join(".node_libraries"));
    }

    dirs
}
```

Reply:
Does the return value have to be `Vec<PathBuf>`? If possible, I would like to return it as `Vec<String>` as before. Other than that point, this implementation is fine.

Reply:
Does the return value have to be Vec<PathBuf>? If possible, I would like to return it as Vec<String> as before. Other than that point, this implementation is fine.

Reply:
Oh right, that should be it. Or even better, `Vec<Rc<str>>` or `Vec<Box<str>>`. Try to avoid String if we're returning immutable strings.

Reply:
That's right, or even return` Vec<Rc<str>>` or `Vec<Box<str>>.` Try to avoid String if we're returning immutable string values.

Reply:
Please wait while I work on this issue some more.